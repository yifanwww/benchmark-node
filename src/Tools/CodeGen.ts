import type { TestFn } from '../types';
import { Hrtime } from '../types.internal';

enum TesterContextEnum {
    TestFn = 'testFn',

    Args = 'args',
    RestArgs = 'restArgs',
    Ops = 'ops',
    Workload = 'workload',

    Setup = 'setup',
    Cleanup = 'cleanup',
}

export interface TesterContext {
    [TesterContextEnum.TestFn]: TestFn;

    [TesterContextEnum.Args]?: ReadonlyArray<unknown>;
    [TesterContextEnum.RestArgs]?: ReadonlyArray<unknown>;
    [TesterContextEnum.Ops]: number;
    [TesterContextEnum.Workload]: boolean;

    [TesterContextEnum.Setup]?: () => void;
    [TesterContextEnum.Cleanup]?: () => void;
}

export type Tester = (context: TesterContext) => { elapsed: Hrtime };

export interface CodeGenArgumentOptions {
    count: number;
    rest?: boolean;
}

export interface CodeGenOptions {
    argument: CodeGenArgumentOptions;
}

export class CodeGen {
    private static cgid: number = 0;

    public static createTester = (options: CodeGenOptions) => new CodeGen(options).createTester();

    private declare readonly id: string;

    private declare readonly argument: CodeGenArgumentOptions;

    public constructor(options: CodeGenOptions) {
        CodeGen.cgid++;
        this.id = CodeGen.cgid.toString();

        this.argument = options.argument;
    }

    private generatePickArguments(): string {
        const code: string[] = [];

        for (let i = 0; i < this.argument.count; i++) {
            // No need to check `context#.${TesterContextEnum.Args}` exists or not.
            code.push(`const arg${i}_# = context#.${TesterContextEnum.Args}[${i}];`);
        }

        if (this.argument.rest) {
            // No need to check `context#.${TesterContextEnum.RestArgs}` exists or not.
            code.push(`const restArg# = context#.${TesterContextEnum.RestArgs};`);
        }

        return code.join('\n');
    }

    private generateTestFnCall(): string {
        const code: string[] = [];

        for (let i = 0; i < this.argument.count; i++) {
            code.push(`arg${i}_#`);
        }

        if (this.argument.rest) {
            code.push('restArg#');
        }

        return `testFn#(${code.join(', ')})`;
    }

    private interpolate = (str: string) => str.replace(/#/g, this.id);

    private removeEmptyLines(str: string) {
        return str
            .split('\n')
            .filter((value) => value !== '')
            .join('\n');
    }

    public createTester(): Tester {
        const body = this.removeEmptyLines(
            `
if (context#.${TesterContextEnum.Setup}) context#.${TesterContextEnum.Setup}();

const testFn# = context#.${TesterContextEnum.TestFn};
const workload# = context#.${TesterContextEnum.Workload};

${this.generatePickArguments()}

let return#;

const begin# = process.hrtime();
for (let i# = 0; i# < context#.${TesterContextEnum.Ops}; i#++) {
    if (workload#) {
        return# = ${this.generateTestFnCall()};
    } else {
        return# = undefined;
    }
}
const elapsed# = process.hrtime(begin#);

if (context#.${TesterContextEnum.Cleanup}) context#.${TesterContextEnum.Cleanup}();

return { elapsed: elapsed#, _internal_return: return# };
`,
        );

        // eslint-disable-next-line @typescript-eslint/no-implied-eval
        return Function(this.interpolate('context#'), this.interpolate(body)) as Tester;
    }
}
