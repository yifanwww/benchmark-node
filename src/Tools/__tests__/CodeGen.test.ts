import { CodeGen } from '../CodeGen';

describe(`Test method \`${CodeGen.name}\``, () => {
    it('create tester with 0 argument', () => {
        const tester = CodeGen.createTester({ argument: { count: 0 } });
        expect(tester.toString()).toMatchSnapshot();
    });

    it('create tester with 1 argument', () => {
        const tester = CodeGen.createTester({ argument: { count: 1 } });
        expect(tester.toString()).toMatchSnapshot();
    });

    it('create tester with 2 arguments', () => {
        const tester = CodeGen.createTester({ argument: { count: 2 } });
        expect(tester.toString()).toMatchSnapshot();
    });

    it('create tester with 3 arguments', () => {
        const tester = CodeGen.createTester({ argument: { count: 3 } });
        expect(tester.toString()).toMatchSnapshot();
    });

    it('create tester with 4 arguments', () => {
        const tester = CodeGen.createTester({ argument: { count: 4 } });
        expect(tester.toString()).toMatchSnapshot();
    });

    it('create tester with rest arguments', () => {
        const tester = CodeGen.createTester({ argument: { count: 0, rest: true } });
        expect(tester.toString()).toMatchSnapshot();
    });

    it('create tester with normal arguments and rest arguments', () => {
        const tester = CodeGen.createTester({ argument: { count: 5, rest: true } });
        expect(tester.toString()).toMatchSnapshot();
    });
});
