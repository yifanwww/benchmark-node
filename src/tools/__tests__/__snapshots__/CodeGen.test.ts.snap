// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test method \`CodeGen\` create tester with 0 argument 1`] = `
"function anonymous(context1
) {
if (context1.setup) context1.setup();
const testFn1 = context1.testFn;
let return1;
const begin1 = process.hrtime();
for (let i1 = 0; i1 < context1.ops; i1++) {
return1 = testFn1();
}
const elapsed1 = process.hrtime(begin1);
if (context1.teardown) context1.teardown();
return { elapsed: elapsed1, _internal_return: return1 };
}"
`;

exports[`test method \`CodeGen\` create tester with 1 argument 1`] = `
"function anonymous(context2
) {
if (context2.setup) context2.setup();
const testFn2 = context2.testFn;
const arg0_2 = context2.args[0];
let return2;
const begin2 = process.hrtime();
for (let i2 = 0; i2 < context2.ops; i2++) {
return2 = testFn2(arg0_2);
}
const elapsed2 = process.hrtime(begin2);
if (context2.teardown) context2.teardown();
return { elapsed: elapsed2, _internal_return: return2 };
}"
`;

exports[`test method \`CodeGen\` create tester with 2 arguments 1`] = `
"function anonymous(context3
) {
if (context3.setup) context3.setup();
const testFn3 = context3.testFn;
const arg0_3 = context3.args[0];
const arg1_3 = context3.args[1];
let return3;
const begin3 = process.hrtime();
for (let i3 = 0; i3 < context3.ops; i3++) {
return3 = testFn3(arg0_3, arg1_3);
}
const elapsed3 = process.hrtime(begin3);
if (context3.teardown) context3.teardown();
return { elapsed: elapsed3, _internal_return: return3 };
}"
`;

exports[`test method \`CodeGen\` create tester with 3 arguments 1`] = `
"function anonymous(context4
) {
if (context4.setup) context4.setup();
const testFn4 = context4.testFn;
const arg0_4 = context4.args[0];
const arg1_4 = context4.args[1];
const arg2_4 = context4.args[2];
let return4;
const begin4 = process.hrtime();
for (let i4 = 0; i4 < context4.ops; i4++) {
return4 = testFn4(arg0_4, arg1_4, arg2_4);
}
const elapsed4 = process.hrtime(begin4);
if (context4.teardown) context4.teardown();
return { elapsed: elapsed4, _internal_return: return4 };
}"
`;

exports[`test method \`CodeGen\` create tester with 4 arguments 1`] = `
"function anonymous(context5
) {
if (context5.setup) context5.setup();
const testFn5 = context5.testFn;
const arg0_5 = context5.args[0];
const arg1_5 = context5.args[1];
const arg2_5 = context5.args[2];
const arg3_5 = context5.args[3];
let return5;
const begin5 = process.hrtime();
for (let i5 = 0; i5 < context5.ops; i5++) {
return5 = testFn5(arg0_5, arg1_5, arg2_5, arg3_5);
}
const elapsed5 = process.hrtime(begin5);
if (context5.teardown) context5.teardown();
return { elapsed: elapsed5, _internal_return: return5 };
}"
`;

exports[`test method \`CodeGen\` create tester with rest arguments 1`] = `
"function anonymous(context6
) {
if (context6.setup) context6.setup();
const testFn6 = context6.testFn;
const restArg6 = context6.restArgs;
let return6;
const begin6 = process.hrtime();
for (let i6 = 0; i6 < context6.ops; i6++) {
return6 = testFn6(restArg6);
}
const elapsed6 = process.hrtime(begin6);
if (context6.teardown) context6.teardown();
return { elapsed: elapsed6, _internal_return: return6 };
}"
`;
